lib/helper.js

```diff
@@ -241,11 +276,76 @@ function getToken(fn, pipeline, jobId) {
     return token;
 }
 
+/**
+ * Gets the executor name from job/pipeline annotations
+ * @param {Object} annotations
+ * @param {Object} pipeline
+ * @param {Object} provider
+ * @returns
+ */
+function getExecutorName(annotations, pipeline, provider) {
+    if (provider) {
+        if (provider.environmentType === ARM_CONTAINER) {
+            return `${provider.executor}-arm64`;
+        }
+
+        return provider.executor;
+    }
+    const executorAnnotation = 'screwdriver.cd/executor';
+    const executorAnnotationBeta = 'beta.screwdriver.cd/executor';
+    const pipelineAnnotations = hoek.reach(pipeline, 'annotations', { default: {} });
+
+    const executorFromAnnotations = annotations ? annotations[executorAnnotation] : annotations[executorAnnotationBeta];
+    const executorFromPipelineAnnotations = pipelineAnnotations
+        ? pipelineAnnotations[executorAnnotation]
+        : pipelineAnnotations[executorAnnotationBeta];
+
+    const executorName = executorFromAnnotations || executorFromPipelineAnnotations;
+
+    return executorName || DEFAULT_KEY;
+}
+
+/**
+ * Get bookend key name based on build cluster, executor and provider
+ * @method getBookendKeyName
+ * @param  {Object}             config
+ * @param  {Object}             config.buildClusterName         Build Cluster Name
+ * @param  {Object}             config.annotations              Annotations for the job
+ * @param  {PipelineModel}      config.pipeline                 Pipeline model
+ * @param  {Provider}           config.provider                 Provider model
+ * @param  {isPipelineUpdate}   from pipeline update
+ * @return {String}             Build cluster name
+ */
+async function getBookendKeyName({ buildClusterName, annotations, pipeline, provider }) {
+    const executor = getExecutorName(annotations, pipeline, provider);
+
+    if (provider) {
+        return `${provider.name}.${provider.buildRegion}.${executor}`;
+    }
+
+    let keyName = buildClusterName;
+
+    if (!buildClusterName) {
+        return keyName;
+    }
+
+    const items = buildClusterName.split('_');
+
+    if (items.length > 1) {
+        keyName = items.join('.');
```

Comment:
if you are going to join by `.` anyway, maybe it's better to change the regex here https://github.com/screwdriver-cd/data-schema/blob/ee91314ef2bb10ef288c86971cfaf361574fb214/models/buildCluster.js#L13-L17, so that sd admins can just follow the naming convention

Reply:
join by `.` is for the bookendKey, I can convert it to object if that makes more sense